### ISSUE TITLE
FLamby integration into Fed-BioMed

### ISSUE THREAD

**Issue by srcansiz** [fedbiomed/fedbiomed#324]
In GitLab by @sayed on Jul 9, 2022, 03:01

**Introduction**

The purpose of this issue is to highlight the possible integration of FLamby into Fed-BioMed, and to describe the current situation / work in progress regarding the automation of this integration. The benefits of it could be a gain in terms of visibility and potential contributors, by bringing ready to use advanced scenarios from cross-silo datasets with natural splits.

**What is FLamby ?**

FLamby is a dataset suite developed for cross-silo Federated Learning with natural partitioning.
The suite has been designed to give the ability to users to run a benchmark on their dataset with natural split, to compare different FL strategies.
As of now, 7 datasets are included in this suite. The way it is working is simple:
Each dataset has its own folder section containing standardized files or folders dedicated to specific things:
- a dataset creation scripts folder, to handle downloading, extracting and preprocessing of the dataset. A configuration file containing the path to the dataset, which is used by the dataloader, is automatically created at end of the execution of this script.
- a metadata folder with a csv file, describing how the subjects from the dataset will be split between train and test
- a file to define the dataloader, which is a federated class, e.g. `FedIXITiny(transform=Compose([...]), center=0, train=True, pooled=False)`
- a file handling model definition
- files defining classes for loss and metric
- a common file containing hyperparameters (e.g. batch size, learning rate, optimizer)

**Accomplished work**

Branch link: https://gitlab.inria.fr/fedbiomed/fedbiomed/-/tree/poc/flamby

- In fedbiomed/node folder:

*cli.py* :arrow_right: We introduce the flamby datasets, giving the possibility to the user to select through the interface one dataset, automatically detected in the flamby package. After selection of a dataset, the user needs to input the index of the desired center (some help being displayed: "Select an id between 0 and X" depending on the flamby dataset and its number of centers), it will create one node corresponding to one center. Logically, these steps need to be repeated depending on the number of centers.

*dataset_manager.py* :arrow_right: It treats all the information that have been input through the CLI, and store them in the node database. We include the support of flamby datasets, and we espacially store the center_id and the path to the folder which contains the federated class of the selected dataset, in a dictionary in dataset_parameters field.

New file: *flamby_utils.py* :arrow_right:
This file especially contains a method `get_flamby_datasets()` to automatically detect datasets in the flamby package.

*round.py* :arrow_right: `_set_training_testing_data_loaders(self)` needs to consider flamby in the creation of `training_data_loader` and `testing_data_loader` through `split_train_and_test_data(self)`.

Another approach has to be defined when dealing with flamby datasets.
As we know, we suppose the user has already launched the flamby dataset scripts, and the particularity of these datasets is that they are not divided into different client folders. 

In these circumstances, we here allow the researcher to ignore the definition of the function `training_data` in the training plan. The DataLoaders are handled internally by the nodes in the case of FLamby, with the help of `_split_train_and_test_data_flamby(self)`. Moreover, the federated dataset class of FLamby is designed in a way that the dataset path is already stored in a configuration file at the end of the execution of the dataset creation scripts. The federated class (dataloader) is made to load the data based on the path in this configuration file. It means that the path we indicate through Fed-BioMed at the creation of the node won't be used (so any path can be set, in all cases it won't affect the program - here, a dummy path is internally defined in *cli.py*).

Some transformation functions can optionally be given through the researcher, but the user needs to consider that all the flamby datasets are already internally handling a transformation through their dataloader (this default transform is the one officially used for the flamby benchmark). It means that it's not something mandatory for running the flamby use cases and the user should check all the transformation functions already defined if he wants to add one.

- In notebooks:

Creation of a notebook which highlights the integration through 5 examples (IXI Tiny, Heart Disease, TCGA BRCA, Synthetic, ISIC datasets): https://gitlab.inria.fr/fedbiomed/fedbiomed/-/blob/poc/flamby/notebooks/flamby-integration-into-fedbiomed.ipynb.

- Environment files

*fedbiomed-node.yaml* now includes `pip -e /path/to/FLamby/package`. To make the examples in the notebook run or to make a flamby dataset available for the nodes, it is necessary to change this path pointing to the local package.
The dataset creation scripts have to be executed from this package, detailed steps are described in a documentation for each dataset.

**Setup**

- To make the integration work at the moment, it is necessary to have a local version of the FLamby package on your computer. As I don't know if I can publicly share the package yet, I can send it by email if anyone would like to test it.

- Replace the path to this local package in fedbiomed-node.yaml or fedbiomed-node-macosx.yaml & fedbiomed-researcher.yaml or fedbiomed-researcher-macosx.yaml with the one on your computer.

- Run the dataset creation scripts of any FLamby dataset you want to make available to Fed-BioMed (the process is detailed in a doc file for each dataset).

- You can now create a node as usual and select FLamby options through the CLI. Some help will be displayed to know how many nodes can be set for each dataset.
